Fixes #30010 -- Add docker-compose support

orf opened the pull request:
Ticket: https://code.djangoproject.com/ticket/30010

This adds docker-compose support to Django. You can run the test suite like so:

`docker-compose run postgres`

or for specific Python/DB versions:

`PYTHON_VERSION=3.7 POSTGRES_VERSION=9.6 docker-compose run postgres`

Some thoughts/notes:

1. We need the `.env` file to be called `.env` and be at the root of the project. See here: https://github.com/docker/compose/issues/4189#issuecomment-274915631

2. ~~The django `.git` directory is *huge*. If we include it in the `.dockerignore` file then we cannot use the `bisect` stuff in `runtests`. If we don't include it then images get significantly bigger, and many builds can eat disk space. *Also*: for some reason the `.git` exclusion in dockerignore is not respected on my machine~~ **Edit: Not relevant anymore**

3. ~~I needed to include docker-specific settings under `docker/settings`. A lot of this feels like it could be de-duplicated, but at the same time it's annoyingly different (different DB hosts, different backends, backend specific config etc)~~

4. I was not sure if we want to include `wait-for-it.sh` or download it at build time. I opted for including it, but we can download it easily enough as well. The reason we need this is because when using `docker-compose run` it does not respect healthchecks, so the database may not be ready by the time the tests are running.

jpic commented on docker/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+RUN for f in ./tests/requirements/*.txt; do pip install -r $f; done && \
+    pip install flake8 flake8-isort selenium
+
+COPY --chown=test:test . .
```
This is an expensive operation on Docker, is it really necessary ?
(We used that kind of technique and got rid of it recently, for the better).

orf replied:
Actually in this case it is not needed at all, as we mount the volumes over it anyway. Nice catch!

It's a pretty expensive operation depending on the number of files, but in most cases it's needed as you want to copy most files into the container. How did you replace it out of interest?

jpic replied:
Actually, we went from `RUN chown` to `COPY --chown` that doesn't have the cost i was thinking about. We still got rid of it of --chown last month though, building the runtime environment as root because the runtime user should not have write access to modify the scripts. We create directories runtime should write into after the USER command in case there is no volume mounted on target directories. We just ["try not to need chown at all"](https://yourlabs.io/oss/crudlfap/commit/e4df870e9c7d56411d127b06a3d6793664a6935a) but that also came with other cmd reordering so I'm not really sure about the significance of the perf impact (given that we're not talking about RUN chown).

So, my comment was a mistake .. it turns out it's indeed not necessary because compose files have volumes ...

Keep up the great work :+1:

jpic commented on docker/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+
+ENV PYTHONPATH "${PYTHONPATH}:/django/:/django/docker/"
+
+VOLUME /django
```
Is it necessary ? docker doesn't ask image permission to mount a volume in a directory. If docker compose says `volumes: [.:/django/]` then it will mount it anyway - whether it has been marked as VOLUME or not in the Dockerfile.

orf replied:
It's just metadata about the expected volumes, I'd like to keep it in the image if possible

timgraham commented:
You can DRY this up by creating a base settings file (Jenkins configuration uses `test_sqlite3.py`) with all the common settings. The other settings files look like:
```
from test_sqlite3 import *  # NOQA

DATABASES = {
   ...
}
```

orf commented:
Thanks @timgraham. I've done that and I've added documentation. Can you give any advice on how you want to structure it? I've split the Quickstart into two sections: one with docker and one without.

I'm a little bit biased, but in my mind running the suite through `docker-compose` is more of a quickstart than the current quickstart we have (especially when it comes to installing system packages!), so I moved this to the top of the documentation.

jpic commented:
Another possibility is to have only one DJANGO_SETTINGS_MODULE and in it switch over a DJANGO_SETTINGS_PROFILE variable. Then you can have only one settings file and still vary the settings on an env var. Maybe worth trying and see if you like that better.

orf commented:
Unfortunately this kills the `isort` and `windows` jenkins jobs, as they create a virtualenv in a directory called `.env`. This PR introduces a `.env` file, who's name cannot be changed :/

jpic commented:
`PYTHON_VERSION=3.7 POSTGRES_VERSION=9.6 docker-compose run postgres` works great here ! No need to maintain a django.lxc container anymore to execute tests :+1: :guitar:

@timgraham is it possible to change .env to .venv maybe in the jobs ?

imomaliev commented on docker-compose.yml:
```diff
@@ -0,0 +1,154 @@
+version: "2.3"
```
Is there any particular reason for using version 2? Why not version 3?

imomaliev commented on docker-compose.yml:
```diff
@@ -0,0 +1,154 @@
+version: "2.3"
```
Do we have a particular reason to use version 2 syntax? Why not version 3?

orf replied:
I'm more familiar with 2. During development there where some things that are removed in 3 that I used to use.

I will see if I can upgrade to 3.

imomaliev commented on docker/Dockerfile:
```diff
@@ -0,0 +1,40 @@
+ARG PYTHON_VERSION
+FROM python:${PYTHON_VERSION}-slim
```
How about using alpine images? Do you think we will have more issues maintaining them?

orf replied:
I'd rather not - developers will be more familiar with the tooling present in the standard images, and size is not that much of a concern. There are also issues with the system packages required, versions etc (especially with geospatial ones).

imomaliev commented on docker/Dockerfile-postgis:
```diff
@@ -0,0 +1,9 @@
+ARG POSTGRES_VERSION
+FROM postgres:${POSTGRES_VERSION}
```
I suggest using https://github.com/appropriate/docker-postgis

orf replied:
I do like the idea of using that. Initially I wanted to try and only use official images, but I'm not sure how much point there is to that.

imomaliev commented on docker/docker_settings/test_mysql.py:
```diff
@@ -0,0 +1,24 @@
+from .test_sqlite import *  # NOQA
```
it is not obvious that `test_sqlite` file is base settings when you just look in file tree, I suggest moving base settings file to `base.py` and also do you think naming test settings files like `test_mysql.py` is redundant ?

orf replied:
I was just following the conventions already in place. I like the idea of `base.py` 👍

imomaliev commented on .env:
```diff
@@ -0,0 +1,5 @@
+PYTHON_VERSION=3.5
```
I suggest using `env.template` file instead of this so that users can modify it locally after doing

```sh
cp env.template .env
```

This is common practice which helps to have predefined `.env` file but changing it won't affect git tree state and allows users to integrate it to their workflow better because there are many tools that use `.env` file and many users have it in their project root already

auvipy commented:
what is blocking this pr?

jkblume commented:
hi,
@auvipy 
think there is an ongoing discussion on the origin project for this PR (https://github.com/orf/django-docker-box/issues/1), they are discussing if this should be merged to django/django or a seperate project under the django github account.

felixxm commented:
@orf IMO we can keep this as a documentation issue (see #11838) after we moved `django-docker-box` into Django. Thanks!
